File: CS50x 2023 Solutions/Week 6/ProbSet 6/credit.py
from cs50 import get_string


def main():
    card = get_string("Number: ")
    if isValid(card):
        type = getType(card)
        print(type)
    else:
        print("INVALID")


def isValid(card):
    digits = [int(digit) for digit in card]

    total = 0

    for i in range(len(digits) - 2, -1, -2):
        num = digits[i] * 2
        total += (num % 10) + int((num / 10))

    for i in range(len(digits) - 1, -1, -2):
        total += digits[i]

    return total % 10 == 0


def getType(card):
    first_two = int(card[:2])
    if len(card) == 15 and (first_two == 34 or first_two == 37):
        return "AMEX"
    elif len(card) == 16 and ((51 <= first_two <= 55)):
        return "MASTERCARD"
    elif (len(card) == 13 or len(card) == 16) and int(card[0]) == 4:
        return "VISA"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()

File: CS50x 2023 Solutions/Week 1/ProbSet 1/credit.c
#include "cs50.h"
#include <stdio.h>

int main(void)
{
    long num = get_long("num: ");

    long num2 = num;

    int mod[16];
    for (int i = 1; i < 17; i++)
    {
        mod[i] = num % 10;
        num /= 10;
    }

    int second = 0;
    for (int j = 1; j < 9; j++)
    {
        mod[j * 2] *= 2;
        mod[j * 2] = (mod[j * 2] % 10) + (mod[j * 2] / 10) % 10;
        second += mod[j * 2] + mod[(j * 2) - 1];
    }

    if (second % 10 == 0 &&
        ((num2 >= 4000000000000 && num2 <= 4999999999999) || (num2 >= 4000000000000000 && num2 <= 4999999999999999) ||
         (num2 >= 340000000000000 && num2 <= 349999999999999) || (num2 >= 370000000000000 && num2 <= 379999999999999) ||
         (num2 >= 5100000000000000 && num2 <= 5599999999999999)))
    {
        if ((num2 >= 4000000000000 && num2 <= 4999999999999) || (num2 >= 4000000000000000 && num2 <= 4999999999999999))
        {
            printf("VISA\n");
        }
        else
        {
            if ((num2 >= 340000000000000 && num2 <= 349999999999999) || (num2 >= 370000000000000 && num2 <= 379999999999999))
            {
                printf("AMEX\n");
            }
            else
            {
                if (num2 >= 5100000000000000 && num2 <= 5599999999999999)
                {
                    printf("MASTERCARD\n");
                }
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}